### PR TITLE
Support row-based Hive UDFs [databricks]

### PR DESCRIPTION
**Pull request by firestarman**
This fixes https://github.com/NVIDIA/spark-rapids/issues/3904

Signed-off-by: Firestarman <firestarmanllc@gmail.com>

### CODE REVIEW

**Comment by firestarman** [comment]
Mark as draft, because needing to try an external app next.

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

File: aggregator/pom.xml [hunk 1]
```diff
@@ -100,9 +100,27 @@
                             <pattern>org.apache.hadoop.hive.</pattern>
                             <shadedPattern>${rapids.shade.package}.hadoop.hive.</shadedPattern>
                             <excludes>
+                                <!-- FIXME
+                                    Seems we need to exclude all the public classes being used
+                                    directly and indirectly by plugin to support row-based Hive
+                                    UDF, otherwise it will get a ClassNotFoundException as below:
+                                        E Caused by: java.lang.ClassNotFoundException: com.nvidia.shaded.spark.hadoop.hive.serde2.objectinspector.ObjectInspector
+                                        E     at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
+                                    Or, just simply remove this relocation.
```
Comment (jlowe) [inline aggregator/pom.xml:109]: We need to shade hive as part of shading ORC (we need to shade it and all of its dependencies, or we risk the hive jar either missing in the Spark distribution at runtime or conflicting with the version that the ORC version we're shading needs).

I think we only need to limit the hive relocations to the hive-storage-api classes.  Unfortunately there isn't a nice package prefix that would select only those classes.  We may need to split up the input/output classes that need ORC (and thus the relocations via shading) from the UDF classes that just need the aggregation part of the shading (i.e.: no relocations).  Separating out the Hive-related UDF classes and/or the ORC input/output classes into a separate Maven project would give us a way to control the shading on a more granular basis.

Comment (firestarman) [reply aggregator/pom.xml]: Thanks Jason. Good to know.
Is this change here OK for now ?

**Comment by firestarman** [comment]
build

File: aggregator/pom.xml (hunk 1 above)
Comment (jlowe) [reply aggregator/pom.xml]: I think what's here is quite fragile, but I guess it's OK for now.  However we should prioritize #1398 to fix this in a better way.

File: sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala [hunk 2]
```diff
@@ -45,59 +45,86 @@ object GpuHiveOverrides {
...
           override def tagExprForGpu(): Unit = {
-            a.function match {
-              case _: RapidsUDF =>
-              case _ =>
-                willNotWorkOnGpu(s"Hive UDF ${a.name} implemented by " +
-                    s"${a.funcWrapper.functionClassName} does not provide a GPU implementation")
+            if (opRapidsFunc.isEmpty && !conf.isCpuBasedUDFEnabled) {
+              willNotWorkOnGpu(s"neither Hive SimpleUDF ${a.name} implemented by " +
+                s"${a.funcWrapper.functionClassName} does not provide a GPU implementation," +
+                s" nor the conf `${RapidsConf.ENABLE_CPU_BASED_UDF.key}` is enabled")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala:64]: I think the `neither` here ends up being a double-negative and confusing.  I suggest dropping it and simply stating what's not provided, e.g.:
```suggestion
              willNotWorkOnGpu(s"Hive SimpleUDF ${a.name} implemented by " +
                s"${a.funcWrapper.functionClassName} does not provide a GPU implementation " +
                s"and CPU-based UDFs are not enabled by `${RapidsConf.ENABLE_CPU_BASED_UDF.key}`")
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala [hunk 3]
```diff
@@ -45,59 +45,86 @@ object GpuHiveOverrides {
...
           override def tagExprForGpu(): Unit = {
-            a.function match {
-              case _: RapidsUDF =>
-              case _ =>
-                willNotWorkOnGpu(s"Hive GenericUDF ${a.name} implemented by " +
-                    s"${a.funcWrapper.functionClassName} does not provide a GPU implementation")
+            if (opRapidsFunc.isEmpty && !conf.isCpuBasedUDFEnabled) {
+              willNotWorkOnGpu(s"neither Hive GenericUDF ${a.name} implemented by " +
+                s"${a.funcWrapper.functionClassName} does not provide a GPU implementation," +
+                s" nor the conf `${RapidsConf.ENABLE_CPU_BASED_UDF.key}` is enabled")
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala:105]: Similar double-negative with neither:
```suggestion
              willNotWorkOnGpu(s"Hive GenericUDF ${a.name} implemented by " +
                s"${a.funcWrapper.functionClassName} does not provide a GPU implementation " +
                s"and CPU-based UDFs are not enabled by `${RapidsConf.ENABLE_CPU_BASED_UDF.key}`")
```

**Review by jlowe — COMMENTED** [review]
Minor suggestion on willNotWork wording but otherwise lgtm.

File: sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala (hunk 2 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala]: Nice catch

File: sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala (hunk 3 above)
Comment (firestarman) [reply sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/GpuHiveOverrides.scala]: Nice catch

**Comment by firestarman** [comment]
build

**Review by jlowe — APPROVED** [review]

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Pull request merged**